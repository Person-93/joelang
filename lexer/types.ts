// token types
export type TokenType = 'paren' | 'bracket' | 'brace' | 'bool' | 'number' | 'string' | 'name' | 'keyword' | 'operator' | 'nil' | 'filepath' | 'separator' | 'comment';

// info about a token
export type Token = {
	type: TokenType;
	start: number; // cursor position of the beginning of this token
	end: number; // cursor position immediately after this token
	value: string;
};

// reserved keywords
export const keywords = ['const', 'let', 'if', 'is', 'in', 'for', 'switch', 'return', 'import', 'from', 'class', 'extends', 'implements', 'this', 'static', 'public', 'private'] as const;
type Keyword = typeof keywords[number];

// operators
export const operators = ['+', '++', '-', '--', '*', '/', '%', '^e', '^^', '|', '||', '&', '&&', '.', '..', '...', '!', ':', ';', ',', '<', '>', '=', '==', '!=', '===', '!==', '<=>'] as const;
type Operator = typeof operators[number];

// special Values
const specialValues = ['true', 'false', 'nil'] as const;
type SpecialValue = typeof specialValues[number];
export const specialValueTypes: Record<SpecialValue, TokenType> = {
	true: 'bool',
	false: 'bool',
	nil: 'nil',
};

// syntax patterns
export const patterns = {
	// single characters
	AMPERSAND: '&',
	ASTERISK: '*',
	BANG: '!',
	CARET: '^',
	COLON: ':',
	COMMA: ',',
	DOUBLE_QUOTE: '"',
	EQUALS: '=',
	FORWARD_SLASH: '/',
	HASH: '#',
	MINUS: '-',
	PIPE: '|',
	PERIOD: '.',
	PLUS: '+',
	QUESTION: '?',
	SEMICOLON: ';',
	SINGLE_QUOTE: "'",
	UNDERSCORE: '_',

	// regexes
	DIGITS: /[0-9]/,
	FILEPATH: /[a-zA-Z0-9-_./]/, // characters in filepath, excluding the front: @
	LETTERS: /[a-z]/i,
	NEWLINE: /(\r\n|\r|\n)/,
	WHITESPACE: /\s/,
};
