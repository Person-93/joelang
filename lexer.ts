import Lexer from "./lexer/lexer";
import LexerError from "./lexer/error";
import fs from 'fs-extra';

void (async (): Promise<void> => {
    try {
		const tokens = new Lexer().lexify(process.argv[2]);

		// filename is 3rd arg
		if (process.argv[3]) {
			await fs.writeFile(process.argv[3], JSON.stringify(tokens, undefined, '\t'));
		} else {
			console.log(tokens);
		}
	} catch (e) {
		const error = e as LexerError;

		console.log(`Error: ${error.message}`);
		console.debug('Extracted tokens:');
		console.table(error.getTokens());

		console.log('Stack Trace:');
		console.error(error.stack);
	}
})();
