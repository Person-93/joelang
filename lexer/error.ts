import { Token } from './types';

/**
 * Custom error class so that we can display the already-extracted tokens
 * which will help the user see where the lexer is up to and got stuck
 */
export default class Error extends TypeError {
	private tokens;

	constructor (message: string, tokens: Token[]) {
		super(message);

		this.tokens = tokens;
	}

	getTokens (): Token[] {
		return this.tokens;
	}
}
