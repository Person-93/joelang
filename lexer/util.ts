export const standardizeLineEndings = (code: string): string => code.replace(/\r\n/g, '\n');
