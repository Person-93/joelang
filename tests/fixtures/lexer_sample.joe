# This is sample .joe file to prove the lexer works.
# Please note this file is not syntatically correct.

import lexer from ./lexer;
import lexer2 from /lexer;

class Foo extends Bar, Baz implements AbstractFooBar, AnotherAbstractClass {
	static bar = 'baz'; // public static property
	private isDone? = true // private instance property

	/** method with no args and no return */
	foo {
		const size = switch someNumber {
			is 1, is 2: 'small';
			is in 3..10: 'medium';
			is 11: 'large';
			else: 'off the charts';
		}

		let myTuple = [
			1,
			10 if this.isDone?, // if false, element will not be in tuple
			3
		];

		let count = 1..myTuple[2] // new Tuple [1, 2, 3]
		let countDown = myTuple[1]..0 // new Tuple [10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
		let partitionedTuple = [
			countDown[:1],
			countDown[2:6],
			countDown[7:],
		]; // new nested Tuple [[10, 9], [8, 7, 6, 5, 4], [3, 2, 1, 0]]
		let selectedTupleItems = countDown[1, 4, 9] // new Tuple [9, 6, 1]

		let myObject = {
			a: 1,
			b: 'coat' if this.isDone?, // if false, property will not be in object
			c: cond ? true : false, // regular ternary - property will always be there, value differs
			d: nil,
		};

		let partialObj = myObject[a, d] // new object {a: 1, d: nil}
	}

	/** with args */
	methodWithArgs (arg1: bool, arg2: Tuple, arg3 = true, ...otherStuff: Thing[]) { ... }

	/** with one return val */
	methodThatReturnsBool? : bool {
		return
	}

	/** multiple return vals */
	methodThatReturnsBool? : bool, number, Thing {
		return false, 5, new Thing
	}

	/** method that throws error */
	careful! { ... }

	/** method with bool return that throws error */
	careful?! : bool { ... }
}
