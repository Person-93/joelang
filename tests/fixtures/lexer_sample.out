[
	{
		"type": "comment",
		"start": 0,
		"end": 52,
		"value": "# This is sample .joe file to prove the lexer works."
	},
	{
		"type": "comment",
		"start": 53,
		"end": 105,
		"value": "# Please note this file is not syntatically correct."
	},
	{
		"type": "keyword",
		"start": 107,
		"end": 113,
		"value": "import"
	},
	{
		"type": "name",
		"start": 114,
		"end": 119,
		"value": "lexer"
	},
	{
		"type": "keyword",
		"start": 120,
		"end": 124,
		"value": "from"
	},
	{
		"type": "filepath",
		"start": 125,
		"end": 132,
		"value": "./lexer"
	},
	{
		"type": "separator",
		"start": 132,
		"end": 133,
		"value": ";"
	},
	{
		"type": "keyword",
		"start": 134,
		"end": 140,
		"value": "import"
	},
	{
		"type": "name",
		"start": 141,
		"end": 147,
		"value": "lexer2"
	},
	{
		"type": "keyword",
		"start": 148,
		"end": 152,
		"value": "from"
	},
	{
		"type": "filepath",
		"start": 153,
		"end": 159,
		"value": "/lexer"
	},
	{
		"type": "separator",
		"start": 159,
		"end": 160,
		"value": ";"
	},
	{
		"type": "keyword",
		"start": 162,
		"end": 167,
		"value": "class"
	},
	{
		"type": "name",
		"start": 168,
		"end": 171,
		"value": "Foo"
	},
	{
		"type": "keyword",
		"start": 172,
		"end": 179,
		"value": "extends"
	},
	{
		"type": "name",
		"start": 180,
		"end": 183,
		"value": "Bar"
	},
	{
		"type": "separator",
		"start": 183,
		"end": 184,
		"value": ","
	},
	{
		"type": "name",
		"start": 185,
		"end": 188,
		"value": "Baz"
	},
	{
		"type": "keyword",
		"start": 189,
		"end": 199,
		"value": "implements"
	},
	{
		"type": "name",
		"start": 200,
		"end": 214,
		"value": "AbstractFooBar"
	},
	{
		"type": "separator",
		"start": 214,
		"end": 215,
		"value": ","
	},
	{
		"type": "name",
		"start": 216,
		"end": 236,
		"value": "AnotherAbstractClass"
	},
	{
		"type": "brace",
		"start": 237,
		"end": 238,
		"value": "{"
	},
	{
		"type": "keyword",
		"start": 240,
		"end": 246,
		"value": "static"
	},
	{
		"type": "name",
		"start": 247,
		"end": 250,
		"value": "bar"
	},
	{
		"type": "operator",
		"start": 251,
		"end": 252,
		"value": "="
	},
	{
		"type": "string",
		"start": 253,
		"end": 258,
		"value": "baz"
	},
	{
		"type": "separator",
		"start": 258,
		"end": 259,
		"value": ";"
	},
	{
		"type": "comment",
		"start": 260,
		"end": 285,
		"value": "// public static property"
	},
	{
		"type": "keyword",
		"start": 287,
		"end": 294,
		"value": "private"
	},
	{
		"type": "name",
		"start": 295,
		"end": 302,
		"value": "isDone?"
	},
	{
		"type": "operator",
		"start": 303,
		"end": 304,
		"value": "="
	},
	{
		"type": "bool",
		"start": 305,
		"end": 309,
		"value": "true"
	},
	{
		"type": "comment",
		"start": 310,
		"end": 338,
		"value": "// private instance property"
	},
	{
		"type": "comment",
		"start": 341,
		"end": 381,
		"value": "/** method with no args and no return */"
	},
	{
		"type": "name",
		"start": 383,
		"end": 386,
		"value": "foo"
	},
	{
		"type": "brace",
		"start": 387,
		"end": 388,
		"value": "{"
	},
	{
		"type": "keyword",
		"start": 391,
		"end": 396,
		"value": "const"
	},
	{
		"type": "name",
		"start": 397,
		"end": 401,
		"value": "size"
	},
	{
		"type": "operator",
		"start": 402,
		"end": 403,
		"value": "="
	},
	{
		"type": "keyword",
		"start": 404,
		"end": 410,
		"value": "switch"
	},
	{
		"type": "name",
		"start": 411,
		"end": 421,
		"value": "someNumber"
	},
	{
		"type": "brace",
		"start": 422,
		"end": 423,
		"value": "{"
	},
	{
		"type": "keyword",
		"start": 427,
		"end": 429,
		"value": "is"
	},
	{
		"type": "number",
		"start": 430,
		"end": 431,
		"value": "1"
	},
	{
		"type": "separator",
		"start": 431,
		"end": 432,
		"value": ","
	},
	{
		"type": "keyword",
		"start": 433,
		"end": 435,
		"value": "is"
	},
	{
		"type": "number",
		"start": 436,
		"end": 437,
		"value": "2"
	},
	{
		"type": "separator",
		"start": 437,
		"end": 438,
		"value": ":"
	},
	{
		"type": "string",
		"start": 439,
		"end": 446,
		"value": "small"
	},
	{
		"type": "separator",
		"start": 446,
		"end": 447,
		"value": ";"
	},
	{
		"type": "keyword",
		"start": 451,
		"end": 453,
		"value": "is"
	},
	{
		"type": "keyword",
		"start": 454,
		"end": 456,
		"value": "in"
	},
	{
		"type": "number",
		"start": 457,
		"end": 458,
		"value": "3"
	},
	{
		"type": "operator",
		"start": 458,
		"end": 460,
		"value": ".."
	},
	{
		"type": "number",
		"start": 460,
		"end": 462,
		"value": "10"
	},
	{
		"type": "separator",
		"start": 462,
		"end": 463,
		"value": ":"
	},
	{
		"type": "string",
		"start": 464,
		"end": 472,
		"value": "medium"
	},
	{
		"type": "separator",
		"start": 472,
		"end": 473,
		"value": ";"
	},
	{
		"type": "keyword",
		"start": 477,
		"end": 479,
		"value": "is"
	},
	{
		"type": "number",
		"start": 480,
		"end": 482,
		"value": "11"
	},
	{
		"type": "separator",
		"start": 482,
		"end": 483,
		"value": ":"
	},
	{
		"type": "string",
		"start": 484,
		"end": 491,
		"value": "large"
	},
	{
		"type": "separator",
		"start": 491,
		"end": 492,
		"value": ";"
	},
	{
		"type": "name",
		"start": 496,
		"end": 500,
		"value": "else"
	},
	{
		"type": "separator",
		"start": 500,
		"end": 501,
		"value": ":"
	},
	{
		"type": "string",
		"start": 502,
		"end": 518,
		"value": "off the charts"
	},
	{
		"type": "separator",
		"start": 518,
		"end": 519,
		"value": ";"
	},
	{
		"type": "brace",
		"start": 522,
		"end": 523,
		"value": "}"
	},
	{
		"type": "keyword",
		"start": 527,
		"end": 530,
		"value": "let"
	},
	{
		"type": "name",
		"start": 531,
		"end": 538,
		"value": "myTuple"
	},
	{
		"type": "operator",
		"start": 539,
		"end": 540,
		"value": "="
	},
	{
		"type": "bracket",
		"start": 541,
		"end": 542,
		"value": "["
	},
	{
		"type": "number",
		"start": 546,
		"end": 547,
		"value": "1"
	},
	{
		"type": "separator",
		"start": 547,
		"end": 548,
		"value": ","
	},
	{
		"type": "number",
		"start": 552,
		"end": 554,
		"value": "10"
	},
	{
		"type": "keyword",
		"start": 555,
		"end": 557,
		"value": "if"
	},
	{
		"type": "keyword",
		"start": 558,
		"end": 562,
		"value": "this"
	},
	{
		"type": "operator",
		"start": 562,
		"end": 563,
		"value": "."
	},
	{
		"type": "name",
		"start": 563,
		"end": 570,
		"value": "isDone?"
	},
	{
		"type": "separator",
		"start": 570,
		"end": 571,
		"value": ","
	},
	{
		"type": "comment",
		"start": 572,
		"end": 613,
		"value": "// if false, element will not be in tuple"
	},
	{
		"type": "number",
		"start": 617,
		"end": 618,
		"value": "3"
	},
	{
		"type": "bracket",
		"start": 621,
		"end": 622,
		"value": "]"
	},
	{
		"type": "separator",
		"start": 622,
		"end": 623,
		"value": ";"
	},
	{
		"type": "keyword",
		"start": 627,
		"end": 630,
		"value": "let"
	},
	{
		"type": "name",
		"start": 631,
		"end": 636,
		"value": "count"
	},
	{
		"type": "operator",
		"start": 637,
		"end": 638,
		"value": "="
	},
	{
		"type": "number",
		"start": 639,
		"end": 640,
		"value": "1"
	},
	{
		"type": "operator",
		"start": 640,
		"end": 642,
		"value": ".."
	},
	{
		"type": "name",
		"start": 642,
		"end": 649,
		"value": "myTuple"
	},
	{
		"type": "bracket",
		"start": 649,
		"end": 650,
		"value": "["
	},
	{
		"type": "number",
		"start": 650,
		"end": 651,
		"value": "2"
	},
	{
		"type": "bracket",
		"start": 651,
		"end": 652,
		"value": "]"
	},
	{
		"type": "comment",
		"start": 653,
		"end": 675,
		"value": "// new Tuple [1, 2, 3]"
	},
	{
		"type": "keyword",
		"start": 678,
		"end": 681,
		"value": "let"
	},
	{
		"type": "name",
		"start": 682,
		"end": 691,
		"value": "countDown"
	},
	{
		"type": "operator",
		"start": 692,
		"end": 693,
		"value": "="
	},
	{
		"type": "name",
		"start": 694,
		"end": 701,
		"value": "myTuple"
	},
	{
		"type": "bracket",
		"start": 701,
		"end": 702,
		"value": "["
	},
	{
		"type": "number",
		"start": 702,
		"end": 703,
		"value": "1"
	},
	{
		"type": "bracket",
		"start": 703,
		"end": 704,
		"value": "]"
	},
	{
		"type": "operator",
		"start": 704,
		"end": 706,
		"value": ".."
	},
	{
		"type": "number",
		"start": 706,
		"end": 707,
		"value": "0"
	},
	{
		"type": "comment",
		"start": 708,
		"end": 755,
		"value": "// new Tuple [10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]"
	},
	{
		"type": "keyword",
		"start": 758,
		"end": 761,
		"value": "let"
	},
	{
		"type": "name",
		"start": 762,
		"end": 778,
		"value": "partitionedTuple"
	},
	{
		"type": "operator",
		"start": 779,
		"end": 780,
		"value": "="
	},
	{
		"type": "bracket",
		"start": 781,
		"end": 782,
		"value": "["
	},
	{
		"type": "name",
		"start": 786,
		"end": 795,
		"value": "countDown"
	},
	{
		"type": "bracket",
		"start": 795,
		"end": 796,
		"value": "["
	},
	{
		"type": "separator",
		"start": 796,
		"end": 797,
		"value": ":"
	},
	{
		"type": "number",
		"start": 797,
		"end": 798,
		"value": "1"
	},
	{
		"type": "bracket",
		"start": 798,
		"end": 799,
		"value": "]"
	},
	{
		"type": "separator",
		"start": 799,
		"end": 800,
		"value": ","
	},
	{
		"type": "name",
		"start": 804,
		"end": 813,
		"value": "countDown"
	},
	{
		"type": "bracket",
		"start": 813,
		"end": 814,
		"value": "["
	},
	{
		"type": "number",
		"start": 814,
		"end": 815,
		"value": "2"
	},
	{
		"type": "separator",
		"start": 815,
		"end": 816,
		"value": ":"
	},
	{
		"type": "number",
		"start": 816,
		"end": 817,
		"value": "6"
	},
	{
		"type": "bracket",
		"start": 817,
		"end": 818,
		"value": "]"
	},
	{
		"type": "separator",
		"start": 818,
		"end": 819,
		"value": ","
	},
	{
		"type": "name",
		"start": 823,
		"end": 832,
		"value": "countDown"
	},
	{
		"type": "bracket",
		"start": 832,
		"end": 833,
		"value": "["
	},
	{
		"type": "number",
		"start": 833,
		"end": 834,
		"value": "7"
	},
	{
		"type": "separator",
		"start": 834,
		"end": 835,
		"value": ":"
	},
	{
		"type": "bracket",
		"start": 835,
		"end": 836,
		"value": "]"
	},
	{
		"type": "separator",
		"start": 836,
		"end": 837,
		"value": ","
	},
	{
		"type": "bracket",
		"start": 840,
		"end": 841,
		"value": "]"
	},
	{
		"type": "separator",
		"start": 841,
		"end": 842,
		"value": ";"
	},
	{
		"type": "comment",
		"start": 843,
		"end": 903,
		"value": "// new nested Tuple [[10, 9], [8, 7, 6, 5, 4], [3, 2, 1, 0]]"
	},
	{
		"type": "keyword",
		"start": 906,
		"end": 909,
		"value": "let"
	},
	{
		"type": "name",
		"start": 910,
		"end": 928,
		"value": "selectedTupleItems"
	},
	{
		"type": "operator",
		"start": 929,
		"end": 930,
		"value": "="
	},
	{
		"type": "name",
		"start": 931,
		"end": 940,
		"value": "countDown"
	},
	{
		"type": "bracket",
		"start": 940,
		"end": 941,
		"value": "["
	},
	{
		"type": "number",
		"start": 941,
		"end": 942,
		"value": "1"
	},
	{
		"type": "separator",
		"start": 942,
		"end": 943,
		"value": ","
	},
	{
		"type": "number",
		"start": 944,
		"end": 945,
		"value": "4"
	},
	{
		"type": "separator",
		"start": 945,
		"end": 946,
		"value": ","
	},
	{
		"type": "number",
		"start": 947,
		"end": 948,
		"value": "9"
	},
	{
		"type": "bracket",
		"start": 948,
		"end": 949,
		"value": "]"
	},
	{
		"type": "comment",
		"start": 950,
		"end": 972,
		"value": "// new Tuple [9, 6, 1]"
	},
	{
		"type": "keyword",
		"start": 976,
		"end": 979,
		"value": "let"
	},
	{
		"type": "name",
		"start": 980,
		"end": 988,
		"value": "myObject"
	},
	{
		"type": "operator",
		"start": 989,
		"end": 990,
		"value": "="
	},
	{
		"type": "brace",
		"start": 991,
		"end": 992,
		"value": "{"
	},
	{
		"type": "name",
		"start": 996,
		"end": 997,
		"value": "a"
	},
	{
		"type": "separator",
		"start": 997,
		"end": 998,
		"value": ":"
	},
	{
		"type": "number",
		"start": 999,
		"end": 1000,
		"value": "1"
	},
	{
		"type": "separator",
		"start": 1000,
		"end": 1001,
		"value": ","
	},
	{
		"type": "name",
		"start": 1005,
		"end": 1006,
		"value": "b"
	},
	{
		"type": "separator",
		"start": 1006,
		"end": 1007,
		"value": ":"
	},
	{
		"type": "string",
		"start": 1008,
		"end": 1014,
		"value": "coat"
	},
	{
		"type": "keyword",
		"start": 1015,
		"end": 1017,
		"value": "if"
	},
	{
		"type": "keyword",
		"start": 1018,
		"end": 1022,
		"value": "this"
	},
	{
		"type": "operator",
		"start": 1022,
		"end": 1023,
		"value": "."
	},
	{
		"type": "name",
		"start": 1023,
		"end": 1030,
		"value": "isDone?"
	},
	{
		"type": "separator",
		"start": 1030,
		"end": 1031,
		"value": ","
	},
	{
		"type": "comment",
		"start": 1032,
		"end": 1075,
		"value": "// if false, property will not be in object"
	},
	{
		"type": "name",
		"start": 1079,
		"end": 1080,
		"value": "c"
	},
	{
		"type": "separator",
		"start": 1080,
		"end": 1081,
		"value": ":"
	},
	{
		"type": "name",
		"start": 1082,
		"end": 1086,
		"value": "cond"
	},
	{
		"type": "operator",
		"start": 1087,
		"end": 1088,
		"value": "?"
	},
	{
		"type": "bool",
		"start": 1089,
		"end": 1093,
		"value": "true"
	},
	{
		"type": "separator",
		"start": 1094,
		"end": 1095,
		"value": ":"
	},
	{
		"type": "bool",
		"start": 1096,
		"end": 1101,
		"value": "false"
	},
	{
		"type": "separator",
		"start": 1101,
		"end": 1102,
		"value": ","
	},
	{
		"type": "comment",
		"start": 1103,
		"end": 1168,
		"value": "// regular ternary - property will always be there, value differs"
	},
	{
		"type": "name",
		"start": 1172,
		"end": 1173,
		"value": "d"
	},
	{
		"type": "separator",
		"start": 1173,
		"end": 1174,
		"value": ":"
	},
	{
		"type": "nil",
		"start": 1175,
		"end": 1178,
		"value": "nil"
	},
	{
		"type": "separator",
		"start": 1178,
		"end": 1179,
		"value": ","
	},
	{
		"type": "brace",
		"start": 1182,
		"end": 1183,
		"value": "}"
	},
	{
		"type": "separator",
		"start": 1183,
		"end": 1184,
		"value": ";"
	},
	{
		"type": "keyword",
		"start": 1188,
		"end": 1191,
		"value": "let"
	},
	{
		"type": "name",
		"start": 1192,
		"end": 1202,
		"value": "partialObj"
	},
	{
		"type": "operator",
		"start": 1203,
		"end": 1204,
		"value": "="
	},
	{
		"type": "name",
		"start": 1205,
		"end": 1213,
		"value": "myObject"
	},
	{
		"type": "bracket",
		"start": 1213,
		"end": 1214,
		"value": "["
	},
	{
		"type": "name",
		"start": 1214,
		"end": 1215,
		"value": "a"
	},
	{
		"type": "separator",
		"start": 1215,
		"end": 1216,
		"value": ","
	},
	{
		"type": "name",
		"start": 1217,
		"end": 1218,
		"value": "d"
	},
	{
		"type": "bracket",
		"start": 1218,
		"end": 1219,
		"value": "]"
	},
	{
		"type": "comment",
		"start": 1220,
		"end": 1248,
		"value": "// new object {a: 1, d: nil}"
	},
	{
		"type": "brace",
		"start": 1250,
		"end": 1251,
		"value": "}"
	},
	{
		"type": "comment",
		"start": 1254,
		"end": 1270,
		"value": "/** with args */"
	},
	{
		"type": "name",
		"start": 1272,
		"end": 1286,
		"value": "methodWithArgs"
	},
	{
		"type": "paren",
		"start": 1287,
		"end": 1288,
		"value": "("
	},
	{
		"type": "name",
		"start": 1288,
		"end": 1292,
		"value": "arg1"
	},
	{
		"type": "separator",
		"start": 1292,
		"end": 1293,
		"value": ":"
	},
	{
		"type": "name",
		"start": 1294,
		"end": 1298,
		"value": "bool"
	},
	{
		"type": "separator",
		"start": 1298,
		"end": 1299,
		"value": ","
	},
	{
		"type": "name",
		"start": 1300,
		"end": 1304,
		"value": "arg2"
	},
	{
		"type": "separator",
		"start": 1304,
		"end": 1305,
		"value": ":"
	},
	{
		"type": "name",
		"start": 1306,
		"end": 1311,
		"value": "Tuple"
	},
	{
		"type": "separator",
		"start": 1311,
		"end": 1312,
		"value": ","
	},
	{
		"type": "name",
		"start": 1313,
		"end": 1317,
		"value": "arg3"
	},
	{
		"type": "operator",
		"start": 1318,
		"end": 1319,
		"value": "="
	},
	{
		"type": "bool",
		"start": 1320,
		"end": 1324,
		"value": "true"
	},
	{
		"type": "separator",
		"start": 1324,
		"end": 1325,
		"value": ","
	},
	{
		"type": "operator",
		"start": 1326,
		"end": 1329,
		"value": "..."
	},
	{
		"type": "name",
		"start": 1329,
		"end": 1339,
		"value": "otherStuff"
	},
	{
		"type": "separator",
		"start": 1339,
		"end": 1340,
		"value": ":"
	},
	{
		"type": "name",
		"start": 1341,
		"end": 1346,
		"value": "Thing"
	},
	{
		"type": "bracket",
		"start": 1346,
		"end": 1347,
		"value": "["
	},
	{
		"type": "bracket",
		"start": 1347,
		"end": 1348,
		"value": "]"
	},
	{
		"type": "paren",
		"start": 1348,
		"end": 1349,
		"value": ")"
	},
	{
		"type": "brace",
		"start": 1350,
		"end": 1351,
		"value": "{"
	},
	{
		"type": "operator",
		"start": 1352,
		"end": 1355,
		"value": "..."
	},
	{
		"type": "brace",
		"start": 1356,
		"end": 1357,
		"value": "}"
	},
	{
		"type": "comment",
		"start": 1360,
		"end": 1386,
		"value": "/** with one return val */"
	},
	{
		"type": "name",
		"start": 1388,
		"end": 1410,
		"value": "methodThatReturnsBool?"
	},
	{
		"type": "separator",
		"start": 1411,
		"end": 1412,
		"value": ":"
	},
	{
		"type": "name",
		"start": 1413,
		"end": 1417,
		"value": "bool"
	},
	{
		"type": "brace",
		"start": 1418,
		"end": 1419,
		"value": "{"
	},
	{
		"type": "keyword",
		"start": 1422,
		"end": 1428,
		"value": "return"
	},
	{
		"type": "brace",
		"start": 1430,
		"end": 1431,
		"value": "}"
	},
	{
		"type": "comment",
		"start": 1434,
		"end": 1461,
		"value": "/** multiple return vals */"
	},
	{
		"type": "name",
		"start": 1463,
		"end": 1485,
		"value": "methodThatReturnsBool?"
	},
	{
		"type": "separator",
		"start": 1486,
		"end": 1487,
		"value": ":"
	},
	{
		"type": "name",
		"start": 1488,
		"end": 1492,
		"value": "bool"
	},
	{
		"type": "separator",
		"start": 1492,
		"end": 1493,
		"value": ","
	},
	{
		"type": "name",
		"start": 1494,
		"end": 1500,
		"value": "number"
	},
	{
		"type": "separator",
		"start": 1500,
		"end": 1501,
		"value": ","
	},
	{
		"type": "name",
		"start": 1502,
		"end": 1507,
		"value": "Thing"
	},
	{
		"type": "brace",
		"start": 1508,
		"end": 1509,
		"value": "{"
	},
	{
		"type": "keyword",
		"start": 1512,
		"end": 1518,
		"value": "return"
	},
	{
		"type": "bool",
		"start": 1519,
		"end": 1524,
		"value": "false"
	},
	{
		"type": "separator",
		"start": 1524,
		"end": 1525,
		"value": ","
	},
	{
		"type": "number",
		"start": 1526,
		"end": 1527,
		"value": "5"
	},
	{
		"type": "separator",
		"start": 1527,
		"end": 1528,
		"value": ","
	},
	{
		"type": "name",
		"start": 1529,
		"end": 1532,
		"value": "new"
	},
	{
		"type": "name",
		"start": 1533,
		"end": 1538,
		"value": "Thing"
	},
	{
		"type": "brace",
		"start": 1540,
		"end": 1541,
		"value": "}"
	},
	{
		"type": "comment",
		"start": 1544,
		"end": 1575,
		"value": "/** method that throws error */"
	},
	{
		"type": "name",
		"start": 1577,
		"end": 1585,
		"value": "careful!"
	},
	{
		"type": "brace",
		"start": 1586,
		"end": 1587,
		"value": "{"
	},
	{
		"type": "operator",
		"start": 1588,
		"end": 1591,
		"value": "..."
	},
	{
		"type": "brace",
		"start": 1592,
		"end": 1593,
		"value": "}"
	},
	{
		"type": "comment",
		"start": 1596,
		"end": 1644,
		"value": "/** method with bool return that throws error */"
	},
	{
		"type": "name",
		"start": 1646,
		"end": 1655,
		"value": "careful?!"
	},
	{
		"type": "separator",
		"start": 1656,
		"end": 1657,
		"value": ":"
	},
	{
		"type": "name",
		"start": 1658,
		"end": 1662,
		"value": "bool"
	},
	{
		"type": "brace",
		"start": 1663,
		"end": 1664,
		"value": "{"
	},
	{
		"type": "operator",
		"start": 1665,
		"end": 1668,
		"value": "..."
	},
	{
		"type": "brace",
		"start": 1669,
		"end": 1670,
		"value": "}"
	},
	{
		"type": "brace",
		"start": 1671,
		"end": 1672,
		"value": "}"
	}
]